c++/src/kj/buffering.h

```diff
@@ -0,0 +1,203 @@
+    KJ_SWITCH_ONEOF(stream) {
+      KJ_CASE_ONEOF(inputStream, AsyncInputStream*) {
+        return readFromStream(*inputStream);
+      }
+      KJ_CASE_ONEOF(ioStream, AsyncIoStream*) {
+        return readFromStream(*ioStream);
+      }
+      KJ_CASE_ONEOF(capStream, AsyncCapabilityStream*) {
+        return readFromStream(*capStream);
+      }
+    }
+    KJ_UNREACHABLE;
+  }
+
+  Promise<void> readFromStream(AsyncInputStream& stream);
+  Promise<void> readFromStream(AsyncCapabilityStream& stream);
+  // Read as many bytes as possible from the provided stream into the buffer
+
+private:
+  EmptyFdBufferRange(FdRangeInputBuffer& parent);
+  friend class FdRangeInputBuffer;
+
+  FdRangeInputBuffer& parent;
+};
+
+class FdRangeInputBuffer {
+  // A buffer of up to two FD ranges that can be filled by reading from a stream.
+  // An FD range is defined as a range of bytes with FDs associated with the last byte in the range.
+  // FDs being associated with the last byte of the range is for the purpose of the buffering logic,
+  // resulting from the rules of SCM_RIGHTS control messages where a read that includes FDs will not
+  // read past the last byte written by the other side of the socket. One can infer from this rule,
+  // that if the other side of the socket sent multiple messages, any FDs received are part of the
+  // last message.
+  //
+  // To handle partially received messages, the caller can provide the remaining bytes necessary to
+  // complete a FD range when performing a read on a non-empty buffer. This ensures the buffer
+  // is always able to associate FDs with the last byte of a range.
+  //
+  // The buffer is emptied by splitting off prefixes of it. If a split-off prefix includes the last
+  // byte of an FD range with FDs, those FDs are also split off.
+  //
+  // The limit of two buffered FD ranges results from the optimization described in `finishRange`.
+  // As `finishRange` allows the buffer to be aware of the end of an FD range, extra bytes can
+  // be read and will be treated as part of a new FD range. Crucially, it's only necessary to
+  // create a new FD range if the existing range already had FDs in it, otherwise it's safe
+  // to extend the existing range.
+  //
+  // It is assumed that if a second FD range is buffered, data stored in the first FD range
+  // is complete (as FDs are associated with the last byte in the range) and will be
+  // consumed from the buffer before any more bytes are read into the buffer.
+
+public:
+  FdRangeInputBuffer(size_t size, size_t maxFds);
+
+  OneOf<EmptyFdBufferRange, Own<SplittableFdBufferRange>> getRange();
+
+private:
+  Array<byte> bytes;
+  Array<AutoCloseFd> fds;
+
+  struct Empty {};
+  struct FdRange {
+    ArrayPtr<byte> bytes;
+    Array<AutoCloseFd> fds;
+  };
+  struct TwoFdRange {
+    FdRange one;
+    FdRange two;
+  };
+
+  OneOf<Empty, FdRange, TwoFdRange> state = Empty{};
```

Comment:
I don't think we care how many FDs are in the range as long as we have an index of each range. I'm picturing a bunch of small messages with a fd each. Would this work with a vector of FdRange instead?